### ISSUE TITLE
JIT will affect the gradient computation of forward mode

### ISSUE THREAD

**Issue by payphon** [pytorch/pytorch#84681]
### 🐛 Describe the bug

JIT will affect the gradient computation of forward mode for some API

For example, if we directly call `jit_fn` with forward mode, it will output the correct gradient:

```py
import torch

def fn(input, target):
    weight = None
    reduction = "mean"
    return torch.nn.functional.multilabel_soft_margin_loss(input, target, weight=weight, reduction=reduction, )

input = torch.tensor([[1., 1.]], dtype=torch.float32)
target = torch.tensor([[2., 2.]], dtype=torch.float32)

inputs1 = (input.clone().requires_grad_(), target.clone().requires_grad_())
inputs2 = (input.clone().requires_grad_(), target.clone().requires_grad_())

jit_fn = torch.jit.trace(fn, (input, target))
print(torch.autograd.functional.jacobian(jit_fn, inputs2, vectorize=True, strategy='forward-mode'))
# [tensor([[-0.6345, -0.6345]], grad_fn=<ReshapeAliasBackward0>), tensor([[-0.5000, -0.5000]], grad_fn=<ReshapeAliasBackward0>)
```

However, if we call `jit_fn(*inputs1)` before the gradient computation in forward mode with `inputs2`, the gradient will be all 0, which is wrong

```py
import torch

def fn(input, target):
    weight = None
    reduction = "mean"
    return torch.nn.functional.multilabel_soft_margin_loss(input, target, weight=weight, reduction=reduction, )

input = torch.tensor([[1., 1.]], dtype=torch.float32)
target = torch.tensor([[2., 2.]], dtype=torch.float32)

inputs1 = (input.clone().requires_grad_(), target.clone().requires_grad_())
inputs2 = (input.clone().requires_grad_(), target.clone().requires_grad_())


jit_fn = torch.jit.trace(fn, (input, target))
jit_fn(*inputs1)
print(torch.autograd.functional.jacobian(jit_fn, inputs2, vectorize=True, strategy='forward-mode'))
# [tensor([[0., 0.]]), tensor([[0., 0.]])]
```

Interestingly, this cannot affect the gradient computation in reverse mode.

Many APIs suffer from this problem, like `lp_pool1d`

### Versions

pytorch: 1.12.1

**Comment by payphon** [comment]
Other APIs like `Softmin, torch.nn.Softsign, torch.nn.Tanhshrink, torch.nn.TripletMarginWithDistanceLoss, RNNCell, torch.nn.LocalResponseNorm` etc also suffer from such issue. I don't know whether their root causes are the same

**Comment by payphon** [comment]
The cause of this issue may be:
1. A sequence of operations
2. the first-time execution of jit function returns the correct grad
3. the second-time execution of jit function returns the **wrong** grad

Thus, the execution of jit function could break down the chain rule of forward mode, which causes it returns all zero as the gradient for the following executions

```
import torch
from torch.autograd.functional import jacobian

def f(a):
    return torch.add(torch.abs(a), torch.tensor(1.0)) # any single operation will have the correct gradient

a = torch.tensor([-1., 2.], device='cuda')

jit_f = torch.jit.script(f)

print(jacobian(f, a.clone().requires_grad_(), vectorize=True, strategy='forward-mode'))

jit_f(a.clone().requires_grad_()) # without this, the grad will be correct
print(jacobian(jit_f, a.clone().requires_grad_(), vectorize=True, strategy='forward-mode'))
```

```
tensor([[-1., -0.],
        [ 0.,  1.]], device='cuda:0', grad_fn=<ReshapeAliasBackward0>)
tensor([[0., 0.],
        [0., 0.]], device='cuda:0')
```

**Comment by janeyx99** [comment]
TorchScript has been deprecated in favor of torch.export as of 2.11. https://pytorch.org/blog/pytorch-2-11-release-blog/ Closing this issue as we are not further developing on it--please ping if you believe this issue was erroneously closed.
